fairscale/nn/misc/flatten_params_wrapper.py

```diff
@@ -217,30 +279,29 @@ def state_dict(self, prefix: str = ..., keep_vars: bool = ...) -> "OrderedDict[s
 
     # Since we have overloads above, we can use Any here.
     def state_dict(self, *args: Any, **kwargs: Any) -> Any:
-        """Return the wrapped module's state_dict (unflattened)."""
+        """Return the wrapped module's state_dict."""
         if self.is_flattened and self._auto_unflatten_state_dict:
-            with self.unflatten_params(recurse=False):
+            # Returns the original version.
+            with self.unflatten_params():
                 return super().state_dict(*args, **kwargs)
         else:
+            # Returns flattened version.
             return super().state_dict(*args, **kwargs)
 
     def flat_state_dict(self, *args: Any, **kwargs: Any) -> Dict[str, Any]:
         """Return the flattened state_dict."""
         assert self.is_flattened
-        with ExitStack() as stack:
-            # tell any nested FlattenParamsWrapper instances not to auto unflatten
-            for module in self.modules():  # includes self
-                if isinstance(module, FlattenParamsWrapper):
-                    stack.enter_context(module._no_auto_unflatten_state_dict())
-            state_dict = self.state_dict(*args, **kwargs)
-        return state_dict
+        with self._no_auto_unflatten_state_dict():
+            return self.state_dict(*args, **kwargs)
 
     @contextmanager
     def _no_auto_unflatten_state_dict(self) -> Generator:
         backup = self._auto_unflatten_state_dict
         self._auto_unflatten_state_dict = False
-        yield
-        self._auto_unflatten_state_dict = backup
+        try:
+            yield
```

Comment:
(for my learning) why do we want try/finally here?

Reply:
I added a comment. The reason I did it is that in case the exception is properly handled by the upper layer (callers) then we need to properly undo the effect in the finally block.